### PR TITLE
Implement Internal DataProviders Service

### PR DESCRIPTION
**Pull request by uakyol**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)

### CODE REVIEW

**Comment by uakyol** [comment]
part of #103

**Review by wangyaopw — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: 0 of 12 files reviewed, 2 unresolved discussions (waiting on @SanjayVas and @uakyol)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/SpannerDataProvidersService.kt, line 46 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MdJ33DS0xWUH0QVyJ_w:-MdJ33DT6J2yaAvxIBI9:b2zcszg) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a4b3a39c4ff58371a7d2d2ac7aa775367703bbe9/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/SpannerDataProvidersService.kt#L46)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
>       DataProviderReader()
>         .readExternalIdOrNull(client.singleUse(), ExternalId(request.externalDataProviderId))
>         ?.dataProvider
>     if (dataProvider == null) {
>       failGrpc(Status.FAILED_PRECONDITION) {
>         "No DataProvider with externalId ${request.externalDataProviderId}"
>       }
>     }
> ```
</details>

is this the same as 

```
return   DataProviderReader()
        .readExternalIdOrNull(client.singleUse(), ExternalId(request.externalDataProviderId))
        ?.dataProvider?:
      failGrpc(Status.FAILED_PRECONDITION) {
        "No DataProvider with externalId ${request.externalDataProviderId}"
      }
```
?
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/DataProvidersServiceTest.kt, line 66 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MdJ3pDJ6f5v9gBzTpxR:-MdJ3pDJ6f5v9gBzTpxS:bj7g5k4) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a4b3a39c4ff58371a7d2d2ac7aa775367703bbe9/src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/DataProvidersServiceTest.kt#L66)):*
> ```Kotlin
>       )
> 
>     assertThat(dataProviderRead).isEqualTo(createdDataProvider)
> ```

can we also assert the dataprovider is equal to an expected proto and allow the reader to see what the message looks like?

**Review by wangyaopw — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: 0 of 12 files reviewed, 1 unresolved discussion (waiting on @oliver-amzn, @SanjayVas, and @uakyol)

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 6 of 12 files at r1, 6 of 6 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 3 unresolved discussions (waiting on @oliver-amzn, @uakyol, and @zachcwc)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/DataProviderReaderLegacy.kt, line 1 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-Me7R85038MTLLl_o_Nk:-Me7R851-Qn9DnJFJcb2:b-588xvp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/0e04f4f4c5991ee084847e71a849bba518382bc1/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/DataProviderReaderLegacy.kt#L1)):*
> ```Kotlin
> // Copyright 2020 The Cross-Media Measurement Authors
> ```

Is there still interesting code that relies on this? We don't need to keep the entire legacy path around. It can be deleted when it's no longer useful (e.g. when there's not code we'd want to reference/modify).

If there is still more to disentangle, I would recommend `LegacyDataProviderReader` to fit other naming.
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateDataProviderLegacy.kt, line 1 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-Me7S8lU10CcAF-QVXZk:-Me7S8lU10CcAF-QVXZl:b1qjvcq) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/0e04f4f4c5991ee084847e71a849bba518382bc1/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateDataProviderLegacy.kt#L1)):*
> ```Kotlin
> // Copyright 2020 The Cross-Media Measurement Authors
> ```

Same here re: whether this could be deleted, otherwise prefer the name `CreateLegacyDataProvider`.

**Review by uakyol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: 9 of 15 files reviewed, 3 unresolved discussions (waiting on @oliver-amzn, @SanjayVas, @wangyaopw, and @zachcwc)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/SpannerDataProvidersService.kt, line 46 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MdJ33DS0xWUH0QVyJ_w:-MdZkcsn9xZAxtDrjwdU:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a4b3a39c4ff58371a7d2d2ac7aa775367703bbe9/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/SpannerDataProvidersService.kt#L46)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

> ```
>       DataProviderReader()
>         .readExternalIdOrNull(client.singleUse(), ExternalId(request.externalDataProviderId))
>         ?.dataProvider
>     if (dataProvider == null) {
>       failGrpc(Status.FAILED_PRECONDITION) {
>         "No DataProvider with externalId ${request.externalDataProviderId}"
>       }
>     }
> ```
is this the same as 

```
return   DataProviderReader()
        .readExternalIdOrNull(client.singleUse(), ExternalId(request.externalDataProviderId))
        ?.dataProvider?:
      failGrpc(Status.FAILED_PRECONDITION) {
        "No DataProvider with externalId ${request.externalDataProviderId}"
      }
```
?
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/DataProvidersServiceTest.kt, line 66 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MdJ3pDJ6f5v9gBzTpxR:-MdZkeOR27SCNmTZ_A4U:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a4b3a39c4ff58371a7d2d2ac7aa775367703bbe9/src/main/kotlin/org/wfanet/measurement/kingdom/service/internal/testing/DataProvidersServiceTest.kt#L66)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

can we also assert the dataprovider is equal to an expected proto and allow the reader to see what the message looks like?
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/DataProviderReaderLegacy.kt, line 1 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-Me7R85038MTLLl_o_Nk:-Me7WvdgE-VM93AEtF5n:b-1ifjpv) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/0e04f4f4c5991ee084847e71a849bba518382bc1/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/DataProviderReaderLegacy.kt#L1)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Is there still interesting code that relies on this? We don't need to keep the entire legacy path around. It can be deleted when it's no longer useful (e.g. when there's not code we'd want to reference/modify).

If there is still more to disentangle, I would recommend `LegacyDataProviderReader` to fit other naming.
</blockquote></details>

There are tests that depend on this. I would rather keep this around until the new path is safely implemented and do a cleanup. Changed the name to LegacyDataProviderReader
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateDataProviderLegacy.kt, line 1 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-Me7S8lU10CcAF-QVXZk:-Me7ctPaC6twap-XSH5E:bjdocap) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/0e04f4f4c5991ee084847e71a849bba518382bc1/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateDataProviderLegacy.kt#L1)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Same here re: whether this could be deleted, otherwise prefer the name `CreateLegacyDataProvider`.
</blockquote></details>

Changed it to CreateLegacyDataProvider

**Review by SanjayVas — APPROVED** [review]
> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 1 unresolved discussion (waiting on @oliver-amzn, @wangyaopw, and @zachcwc)

**Review by oliver-amzn — COMMENTED** [review]
> Reviewed 2 of 12 files at r1, 6 of 6 files at r2, 1 of 1 files at r3, 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 3 unresolved discussions (waiting on @uakyol, @wangyaopw, and @zachcwc)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/LegacyDataProviderReader.kt, line 1 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MeCEayCEXK4lvRywnyx:-MeCEayCEXK4lvRywnyy:bxqvv2v) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/5a323265c5f5efbb0d2bc7fecd10e8627fe1da71/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/LegacyDataProviderReader.kt#L1)):*
> ```Kotlin
> // Copyright 2020 The Cross-Media Measurement Authors
> ```

nit: 2021
___
*[src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateLegacyDataProviderTest.kt, line 33 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MeCFJG4ERyBDZqyY-8I:-MeCFJG4ERyBDZqyY-8J:b-zagozz) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/5a323265c5f5efbb0d2bc7fecd10e8627fe1da71/src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateLegacyDataProviderTest.kt#L33)):*
> ```Kotlin
> 
> @RunWith(JUnit4::class)
> class CreateLegacyDataProviderTest : KingdomDatabaseTestBase() {
> ```

Question: This renames the current CreateDataProviderTest to Legacy... but shouldn't we also have a test for the newly written Create and Read classes? Or is this covered sufficiently by the service test above?

**Review by SanjayVas — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 2 unresolved discussions (waiting on @oliver-amzn, @wangyaopw, and @zachcwc)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/LegacyDataProviderReader.kt, line 1 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MeCEayCEXK4lvRywnyx:-MeCHteyCIK7qvD_yrmw:b12w2wp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/5a323265c5f5efbb0d2bc7fecd10e8627fe1da71/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/LegacyDataProviderReader.kt#L1)):*
<details><summary><i>Previously, oliver-amzn (Oliver) wrote…</i></summary><blockquote>

nit: 2021
</blockquote></details>

This is a rename (though sadly Git doesn't detect those well by default) so 2020 is correct.
___
*[src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateLegacyDataProviderTest.kt, line 33 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MeCFJG4ERyBDZqyY-8I:-MeCHyCCCyXMGq2ubbEh:bpo4gxc) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/5a323265c5f5efbb0d2bc7fecd10e8627fe1da71/src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateLegacyDataProviderTest.kt#L33)):*
<details><summary><i>Previously, oliver-amzn (Oliver) wrote…</i></summary><blockquote>

Question: This renames the current CreateDataProviderTest to Legacy... but shouldn't we also have a test for the newly written Create and Read classes? Or is this covered sufficiently by the service test above? 
</blockquote></details>

See discussion of implementation detail classes in https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/123#pullrequestreview-693116347 and https://testing.googleblog.com/2015/01/testing-on-toilet-prefer-testing-public.html.

tl;dr these are implementation detail classes that don't need their own tests.

**Review by oliver-amzn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 1 unresolved discussion (waiting on @wangyaopw and @zachcwc)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/LegacyDataProviderReader.kt, line 1 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MeCEayCEXK4lvRywnyx:-MeOpaDJ1sr9lYNRa4MT:b-il964q) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/5a323265c5f5efbb0d2bc7fecd10e8627fe1da71/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/readers/LegacyDataProviderReader.kt#L1)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

This is a rename (though sadly Git doesn't detect those well by default) so 2020 is correct.
</blockquote></details>

Ugh, you're right, missed that. Thx
___
*[src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateLegacyDataProviderTest.kt, line 33 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127#-MeCFJG4ERyBDZqyY-8I:-MeOpUit4WymWNVv45t0:brxp9sq) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/5a323265c5f5efbb0d2bc7fecd10e8627fe1da71/src/test/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/writers/CreateLegacyDataProviderTest.kt#L33)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

See discussion of implementation detail classes in https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/123#pullrequestreview-693116347 and https://testing.googleblog.com/2015/01/testing-on-toilet-prefer-testing-public.html.

tl;dr these are implementation detail classes that don't need their own tests.
</blockquote></details>

Sounds good.

**Review by zachcwc — APPROVED** [review]
> Reviewed 2 of 12 files at r1, 6 of 6 files at r2, 1 of 1 files at r3, 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 1 unresolved discussion (waiting on @wangyaopw)

**Review by oliver-amzn — APPROVED** [review]
> Reviewed 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/127)* status: all files reviewed, 1 unresolved discussion (waiting on @wangyaopw)
